### PR TITLE
Upgrade typescript to 2.5.3

### PR DESCRIPTION
**Pull request by dustinblackman**
## Overview

Upgrades Typescript to 2.5.3. Closes https://github.com/TypeStrong/typedoc/issues/624, https://github.com/TypeStrong/typedoc/pull/586, and https://github.com/TypeStrong/typedoc/pull/573 (as it's gone stale).

### CODE REVIEW

**Comment by kjdelisle** [comment]
@dustinblackman Do you know when this PR will land? This is currently affecting our ability to generate documentation since typescript@2.4.1 throws compiler errors for some of our code that's valid in typescript@2.5.3

**Comment by dustinblackman** [comment]
@kjdelisle It's not waiting on me. Waiting for a maintainer to look at it.

File: src/lib/application.ts [hunk 1]
```diff
@@ -91,8 +91,8 @@ export class Application extends ChildableComponent<Application, AbstractCompone
         super(null);
 
         this.logger    = new ConsoleLogger();
-        this.converter = this.addComponent('converter', Converter);
-        this.renderer  = this.addComponent('renderer', Renderer);
+        this.converter = this.addComponent<Converter>('converter', Converter);
+        this.renderer  = this.addComponent<Renderer>('renderer', Renderer);
```
Comment (aciccarello) [inline src/lib/application.ts]: I'm curious why these components need to be explicitly typed but the PluginHost and Options do not

Comment (dustinblackman) [reply src/lib/application.ts]: It was a change done in 2.5 that I wasn't able to pin down. I'm not familiar enough with the Typedoc codebase.

Comment (aciccarello) [reply src/lib/application.ts]: Looks like the error is

> Type 'AbstractComponent<Application> & Component' is not assignable to type 'Converter'.
  Property 'name' is missing in type 'AbstractComponent<Application> & Component'.

So I think the error is due to a change in TypeScript type inference. I didn't dig deep enough into the class inheritance to figure out why the other components are inferred/casted differently.

**Comment by dustinblackman** [comment]
@aciccarello Using `ts.BindingElement[]` causes the following error with 2.5.3. The only way I found to have it pass and react the same way was with either `ReadonlyArray<ts.BindingElement>` or `ts.NodeArray<ts.BindingElement>`. I'll gladly update the PR if you have something else in the mind, but this was as far as I was able to go.

```
src/lib/converter/nodes/variable-statement.ts|45 col 10 error| TS2352: Type 'NodeArray<BindingElement> | NodeArray<ArrayBindingElement>' cannot be converted to type 'BindingElement[]'. Type 'NodeArray<ArrayBindingElement>' is not comparable to type 'BindingElement[]'. Property 'push' is missing in type 'NodeArray<ArrayBindingElement>'.
```

**Comment by aciccarello** [comment]
Oops, I believe you are right. I think I didn't install the updated version of TypeScript after checking out the code. If it needs to be changed I think I'll go ahead and change the type to the NodeArray since that is a more accurate type.

**Comment by dustinblackman** [comment]
@aciccarello The mistake of being on Github/coding late at night resulted in my linter not working as expected, `NodeArray` doesn't seem to pass at the moment. I'll see if I can figure out before the week ends.

**Comment by dustinblackman** [comment]
@aciccarello Done, looks like were all green! :)

**Comment by aciccarello** [comment]
@dustinblackman Thanks for your help!

**Comment by bajtos** [comment]
@aciccarello thank you for landing this patch. Is there any estimate on when the next version will be published to npmjs.org?

**Comment by aciccarello** [comment]
@bajtos I don't have access/know how to release so I can give you an answer to that. However you should be able to install the got repo to npm if you need this immediately. 

There are quite a few PRs that need to be reviewed too so help looking at those is appreciated to get features out quickly.

**Comment by octatone** [comment]
What is the ETA for pushing this to NPM?

**Pull request merged**